File: Assignment/Services/ASTService.cs
using Antlr4.Runtime;
using Assignment.Abstraction;
using Assignment.Grammar;
using Assignment.Implementation;
using Assignment.Implementation.Visitors;
using Assignment.Services.DOT;
using System;
using System.IO;
using System.Text;

namespace Assignment.Services
{
    public static class ASTService
    {
        public static INode CompileToAST(string source)
        {
            byte[] bytes = Encoding.ASCII.GetBytes(source);
            var lexer = new LanguageLexer(new AntlrInputStream(new MemoryStream(bytes)));
            var tokens = new CommonTokenStream(lexer);
            var parser = new LanguageParser(tokens);

            try
            {
                return new LanguageVisitor().VisitCompileUnit(parser.compileUnit());
            }
            catch (Exception e)
            {
                throw e;
            }
        }

        public static string CompileToPostFix(INode tree)
        {
            try
            {
                return new ForthVisitor().Visit(tree);
            }
            catch (Exception e)
            {
                throw e;
            }
        }

        public static string VisualiseTree(INode tree, string indent = "")
        {
            return new VisualizationVisitor().Visit(tree, indent);
        }

        public static string CompileToDOT(INode tree)
        {
            return new DOTBuilder().Build(tree, "G");
        }
    }
}

File: src/main/java/ce305/services/ASTService.java
package ce305.services;

import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.Lexer;
import org.antlr.v4.runtime.TokenStream;

import ce305.abstraction.INode;
import ce305.abstraction.expressions.ProgramNode;
import ce305.gen.*;
import ce305.implementation.visitors.ASTSorter;
import ce305.implementation.visitors.DotVisitor;
import ce305.implementation.visitors.ParseVisitor;
import ce305.implementation.visitors.PythonVisitor;
import ce305.implementation.visitors.SemanticAnalyser;
import ce305.implementation.visitors.VisualizationVisitor;

public class ASTService {
    public static INode compileToAST(String source) {
        Lexer lexer = new LanguageLexer(CharStreams.fromString(source));
        TokenStream tokenStream = new CommonTokenStream(lexer);
        LanguageParser parser = new LanguageParser(tokenStream);

        // Generate AST
        ProgramNode tree = (ProgramNode)new ParseVisitor().visitCompileUnit(parser.compileUnit());

        // Do Semantics
        tree = new SemanticAnalyser().analyse(tree);

        //Re-order AST so that functions declarations are at the start of each body
        tree = (ProgramNode)new ASTSorter(tree).visit(tree);

        return tree;
    }

    public static String compileToPython(INode tree) {
        String result = new PythonVisitor().visit(tree);
        return result;
    }

    public static String visualizeTree(INode tree) {
        String result = new VisualizationVisitor().visit(tree);
        return result;
    }

    public static String compileToDot(INode tree) {
        String result = new DotVisitor().visit(tree);
        return result;
    }
}